cts/cts-regression.in

```diff
@@ -1,230 +1,257 @@
+
+    # sysexits.h
+    USAGE = 64
+    DATAERR = 65
+    NOINPUT = 66
+    NOUSER = 67
+    NOHOST = 68
+    UNAVAILABLE = 69
+    SOFTWARE = 70
+    OSERR = 71
+    OSFILE = 72
+    CANTCREAT = 73
+    IOERR = 74
+    TEMPFAIL = 75
+    PROTOCOL = 76
+    NOPERM = 77
+    CONFIG = 78
+
+    # Custom
+    FATAL = 100
+    PANIC = 101
+    DISCONNECT = 102
+    OLD = 103
+    DIGEST = 104
+    NOSUCH = 105
+    QUORUM = 106
+    UNSAFE = 107
+    EXISTS = 108
+    MULTIPLE = 109
+    EXPIRED = 110
+    NOT_YET_IN_EFFECT = 111
+    INDETERMINATE = 112
+    UNSATISFIED = 113
+
+    # Other
+    TIMEOUT = 124
+
+    # OCF Resource Agent API 1.1
+    DEGRADED = 190
+    DEGRADED_PROMOTED = 191
+
+    # Custom
+    NONE = 193
+
+    MAX = 255
+
+
+class TestsArgAction(argparse.Action):
+    """Handle special cases for `tests` arguments.
+
+    Enforce a default value of ['cli', 'scheduler'].
+    Replace the 'all' alias with the tests that it represents.
+    Get rid of duplicates.
+
+    The main motivation is that when the `choices` argument of
+    :meth:`parser.add_argument()` is specified, the `default` argument
+    must contain exactly one value (not `None` and not a list). We want
+    our default to be a list of tests, namely `cli` and `scheduler`.
+    """
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        all_tests = ['cli', 'exec', 'fencing', 'scheduler']
+        default_tests = ['cli', 'scheduler']
+
+        if not values:
+            setattr(namespace, self.dest, default_tests)
+            return
+
+        # If no argument is specified, the default gets passed as a
+        # string 'default' instead of as a list ['default']. Probably
+        # a bug in argparse. The below gives us a list.
+        if not isinstance(values, list):
+            values = [values]
+
+        tests = set(values)
+
+        # If 'all', is found, replace it with the tests it represents.
+        try:
+            tests.remove('all')
+            tests.update(set(all_tests))
+        except KeyError:
+            pass
+
+        # Same for 'default'
+        try:
+            tests.remove('default')
+            tests.update(set(default_tests))
+        except KeyError:
+            pass
+
+        setattr(namespace, self.dest, sorted(list(tests)))
+
+
+def error_print(msg):
+    """Print an error message.
+
+    :param msg: Message to print
+    :type msg: str
+    """
+    print('      * ERROR:   %s' % msg)
+
+
+def run_test(test, test_home, verbose=False, valgrind=False):
+    """Run a regression test and return the result.
+
+    :param test: The name of the test to run
+    :type test: str
+    :return: The exit code of the test
+    :rtype: :class:`CrmExit`
+    """
+    print('Executing the %s regression tests' % test)
+    print('=' * 60)
+    if test == 'pacemaker_remote':
+        cmd = [os.path.join(test_home, 'cts-exec'), '-R']
+    else:
+        cmd = [os.path.join(test_home, 'cts-%s' % test)]
```

Comment:
Looking ahead to when all the tests will eventually be modularized, I think it will be worthwhile to define a regression test component class. The constructor can define this command as a member or method return.

The hope for the eventual design is that they can all share infrastructure code and the only real difference between them is a list of specific tests. I'm thinking something like an individual test class, a test group of some sort (equivalent to cts-cli's dates tools crm_mon etc.), and a component class (equivalent to cli scheduler etc.).

So we'd be starting at the top with a component class.

Reply:
Done (`Component` only so far).